src/python/WMCore/WMBS/MySQL/Workflow/GetDeletedBlocksByWorkflow.py

```diff
@@ -0,0 +1,154 @@
+          'deleted': 0,
+          'name': 'Express_Run351572_StreamExpressCosmics_Tier0_REPLAY_2022_ID220531142559_v425_220531_1426'},
+        {'blockname': '/StreamExpressCosmics/Tier0_REPLAY_2022-SiStripCalZeroBias-Express-v425/ALCARECO#4cb4a954-a96c-43b8-9f2e-2ec5f92c06dd',
+          'deleted': 1,
+          'name': 'Express_Run351572_StreamExpressCosmics_Tier0_REPLAY_2022_ID220531142559_v425_220531_1426'},
+         {'blockname': '/StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#d2819fcf-bac2-4fbb-80cf-11b5bc9eeb00',
+          'deleted': 0,
+          'name': 'Express_Run351572_StreamExpressCosmics_Tier0_REPLAY_2022_ID220531142559_v425_220531_1426'}
+        ...
+        ]
+
+        NOTE:
+         * The number of records per workflow and block returned (i.e. number of records per group in the GROUP BY statement)
+           from the query is not related to either the number of blocks nor to the number of workflows, but rather to the
+           combination of number of files in the block and some other factor which increases the granularity (it seems to be
+           the number of records in dbsbuffer_workflow table per file aggregated by workflow), and NO `DISTINCT` requirement
+           in the SELECT statement is needed because we already have them properly grouped.
+         * Once deleted we should NOT expect duplicate records with two different values of the deleted
+           flag to be returned for a single block but we should still create the list of deleted and
+           NotDeleted blocks as sets and eventually create their proper intersection for double check.
+
+        This list needs to be further aggregated by name to produce an aggregated structure per workflow like:
+
+        [{'name': 'Express_Run351572_StreamExpressCosmics_Tier0_REPLAY_2022_ID220531142559_v425_220531_1426'
+          'blocksNotDeleted': ['/StreamExpressCosmics/Tier0_REPLAY_2022-SiStripCalZeroBias-Express-v425/ALCARECO#4cb4a954-a96c-43b8-9f2e-2ec5f92c06dd',
+                               /StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#d2819fcf-bac2-4fbb-80cf-11b5bc9eeb00']
+          'blocksDeleted': [/StreamExpressCosmics/Tier0_REPLAY_2022-SiStripCalZeroBias-Express-v425/ALCARECO#4cb4a954-a96c-43b8-9f2e-2ec5f92c06dd']
+         },
+        ...
+        ]
+
+        :param result: The result as returned by the mysql query execution.
+        :return:       List of dictionaries
+        """
+
+        # First reformat the output in a list of dictionaries per DB record
+        dictResults = DBFormatter.formatDict(self, result)
+
+        # Now aggregate all blocks per workflow:
+        aggResults = []
+        aggResults.append(self._createAggRecord(dictResults.pop()))
+        while dictResults:
+            aggRecord = self._createAggRecord(dictResults.pop())
+            if aggRecord['name'] not in [wf['name'] for wf in aggResults]:
```

Comment:
this is very excessing and unnecessary check. Instead, construct the list of workflow names and put them in a dict. This should be done  before `while` loop. Then, over here you'll only check presence of workflow name in such dict and avoid constant looping of `aggResults`

Reply:
Yeees ... I know this is kind of expensive here. And I hope I understand your suggestion correctly to build a set of all workflow names out of the whole query result and wile aggregating to go and check if a workflow name does exist there. If I did not get it right, then please ignore my comment bellow. Your suggestion could work, but this is not as obvious as it looks at a first glance. What is happening here is:
 * first put a single element in the `aggResults` list
 * then continuously check for every next element/record from the query results if the current one from the cycle is already present in the `aggResults` list or not:
    * If not - add the new one as it is 
    * if present - merge the new one with the already existing one.
    
Before the while cycle the, `aggResults` list is having only a single element so not usefull to create such a list with workflow names out of it at this stage. At the same time if I try to create it (again in advance) from the the set of all names from the query result, then it does not hold any information on if a single record is a new record for the `aggResults`  or has to be aggregated with another one. That's why the status of the `aggResults` list needs to be refreshed on every step. To achieve the same with an external data structure holding workflow names, we should be having a workflow name as a key and a set of flags like `present: True/False` or/and `merged: True/False` telling what the status of the current record in the cycle is in comparison to the  already passed so that the current action to be decided properly  - either merge or add.  So we will more or less  duplicate the whole structure in the expense of increased memory footprint. Yes the search complexity then indeed would be O(1) but as usual this is a trade off between Memory footprint vs CPU cycles. Given that this is a DAO intended to be used for debugging only and not in the code I think the current implementation is good enough.

Reply:
Without giving it much thought. I agree with Valentin, this is likely something that could be accomplished with 5 lines or so. The current implementation makes it look overly complex.

Reply:
And here is one concrete suggestion (there might be a problem here or there though):
```
results = {}
for record in dictResults:
    wfName = record['name']
    results.setdefault(wfName, {'name': wfName, 'blocksDeleted': [], 'blocksNotDeleted': []})
    if record['deleted']:
        results[wfName]['blocksDeleted'].append(record['blockname'])
    else:
        results[wfName]['blocksNotDeleted'].append(record['blockname'])
return results.values()
```

Reply:
That's a good way of doing the job. Thanks @amaltaro

Reply:
done